vendor/bat-native-ledger/src/bat/ledger/internal/core/async_result.h

```diff
@@ -0,0 +1,122 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_ASYNC_RESULT_H_
+#define BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_ASYNC_RESULT_H_
+
+#include <list>
+#include <memory>
+#include <utility>
+
+#include "base/callback.h"
+#include "base/optional.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+
+namespace ledger {
+
+// Represents the result of an asynchronous operation.
+//
+// Example:
+//   AsyncResult<int>::Resolver resolver;
+//   resolver.Complete(42);
+//
+//   AsyncResult<int> result = resolver.result();
+//   result.Listen(base::BindOnce([](const int& value) {}));
+//
+// Listeners are called on the current SequencedTaskRunner, and are guaranteed
+// to be called asynchronously. AsyncResult and Resolver objects are internally
+// reference counted and can be passed between sequences; the internal data
+// structures are updated on the sequence that created the Resolver.
+template <typename T>
+class AsyncResult {
+ public:
+  using CompleteType = T;
+  using CompleteCallback = base::OnceCallback<void(const T&)>;
+
+  void Listen(CompleteCallback on_complete) {
+    Listener listener = {.on_complete = std::move(on_complete),
+                         .task_runner = base::SequencedTaskRunnerHandle::Get()};
+
+    task_runner_->PostTask(FROM_HERE, base::BindOnce(AddListenerInTask, store_,
+                                                     std::move(listener)));
+  }
+
+  class Resolver {
+   public:
+    Resolver() {}
+    void Complete(T&& value) { result_.Complete(std::move(value)); }
+    AsyncResult result() const { return result_; }
+
+   private:
+    AsyncResult result_;
+  };
+
+ private:
+  AsyncResult()
+      : store_(new Store()),
+        task_runner_(base::SequencedTaskRunnerHandle::Get()) {}
+
+  enum class State { kPending, kComplete };
+
+  struct Listener {
+    CompleteCallback on_complete;
+    scoped_refptr<base::SequencedTaskRunner> task_runner;
+  };
+
+  struct Store {
+    Store() {}
+    State state = State::kPending;
+    base::Optional<T> value;
+    std::list<Listener> listeners;
```

Comment:
Any particular reason to use `std::list` vs `std::vector`?

Reply:
Since we're always adding to the end, and only forward-iterating, a linked list is probably most appropriate. (A vector has to be reallocated if it reaches capacity.)

Reply:
Makes sense.